### DISCUSSION TITLE
NYXI Chaos Pro controller, Switch pro Controller Wireless for Nintendo SP5248LED to drive Create3/RPI 4B

### DISCUSSION THREAD

**Discussion by alain-lnpcrd** [iRobotEducation/create3_docs#507] [Questions]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

laptop amd64 ubuntu 22.04

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.3

### Which RMW is your robot running?

FastRTPS

### Does your robot have an assigned namespace? If so, please share.

no

### Is the robot connected to a network? If so, what is the network type?

home

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

None

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

Hi,
I don't know anything about the joystick and I would like to configure my joystick (for switch) to drive the create3 to map  my house using lidar.
Unfortunately the only thing I get is to rotate the robot right or left by pressing the trigger R and moving the right stick and unfortunately neither forward nor backward.
What should I modify and with what
Thanks.

**Comment by slowrunner** [comment]
- Is your robot a TurtleBot4?  (If so, the [Turtlebot4 discussions](https://github.com/turtlebot/turtlebot4/discussions) is a better place to ask this)
- Creating and installing a nyxichaos.config.yaml file for a new controller, and launching the teleop_twist_joy node with the new config.yaml file is a bit involved.  
  - The first step is to find out what codes your controller emits for every button and joystick:  I did this for my "non-standard controller" with the jstest program
  - Then you must create your nyxichaos.config.yaml file to map your codes to the basic operations
  - Next your config file needs to be put somewhere where the launch can find it - /opt/ros/humble/share/teleop_twist_joy/config/ is where all the default files are located - you should not be messing with that location...  you need to find out where yours should go.
  - Then you need to find where your teleop_twist_joy node is being launched and give it the new config file name: for example if launching from the CLI:  (if inside a launch file - I haven't a clue where or what that looks like)
```
ros2 launch teleop_twist_joy teleop-launch.py joy_config:="nyxichaos" 
```

Good luck.

**Reply by alain-lnpcrd** [reply]
Hi,
No it's not a turtlebot4 even if my goal is to run it with the turtlebot4 examples but for now I have just acquired a lidar and a joystick.
Given the price difference between the create3 and the turtlebot4, that it seems not justified by the additional equipment, I tell myself that I have little chance of getting answers, but I haven't tried.
I will make the most of your indications.

**Reply by shamlian** [reply]
The method in the second bullet seems right to me, as well; I feel like it should be easier to get different joystick configs to teleop_twist_joy (a command line option, perhaps, to set axes?), but it's just not. Probably my counterpart for that repo would tell me that PRs are welcome! 😆

**Reply by slowrunner** [reply]
This is my [SNES gamepad config file](https://github.com/slowrunner/wali_pi5/blob/main/configs/docker/snes_slow.config.yaml)

This is my [Logitech F710 config file](https://github.com/slowrunner/wali_pi5/blob/main/configs/docker/F710.config.yaml)

My [joystick](https://github.com/slowrunner/wali_pi5/tree/main/systests/joy) readme

**Comment by alain-lnpcrd** [comment]
Hi
Thanks for you help.
I made progress this morning and by going to github, package and section iRobotEducation/create3_examples/create3_teleop, I saw that I was not using the planned launch of nodes but in particular a node which did not give the same configuration
So, the result with the launch indicated :

```
moi@pi4b-8go-u-22-04:~/create3_ws$ ros2 launch create3_teleop teleop_joystick_launch.py
[INFO] [launch]: All log files can be found below /home/moi/.ros/log/2024-01-11-12-09-58-205401-pi4b-8go-u-22-04-3434
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [joy_node-1]: process started with pid [3435]
[INFO] [teleop_node-2]: process started with pid [3437]
[teleop_node-2] [INFO] [1704971398.816463890] [TeleopTwistJoy]: Teleop enable button 2.
[teleop_node-2] [INFO] [1704971398.816667462] [TeleopTwistJoy]: Turbo on button 5.
[teleop_node-2] [INFO] [1704971398.816694647] [TeleopTwistJoy]: Linear axis x on 1 at scale 0.700000.
[teleop_node-2] [INFO] [1704971398.816729091] [TeleopTwistJoy]: Turbo for linear axis x is scale 1.500000.
[teleop_node-2] [INFO] [1704971398.816749368] [TeleopTwistJoy]: Angular axis yaw on 0 at scale 0.400000.
[teleop_node-2] [INFO] [1704971398.816767850] [TeleopTwistJoy]: Turbo for angular axis yaw is scale 1.000000.
[joy_node-1] [INFO] [1704973754.917239453] [joy_node]: No haptic (rumble) available, skipping initialization
[joy_node-1] [INFO] [1704973754.917586246] [joy_node]: Opened joystick: Pro Controller.  deadzone: 0.300000

```
The joystick communicates with the Pi and by pressing the R trigger (button 5) while maneuvering the left thumb i drive the create3 where i want.
I don't understand what button 2 marked B on the joystick (XYAB group)
If I press the home button "home" for too long, the entry becomes /dev/input/event4 instead of /dev/input/js0 and there the only solution found is a reboot.
The good thing is that I immediately found a solution to be able to direct the create3 map my house more practical than a keyboard.
I am far from understanding the operating process of a joystick.
But knowing that walking the create3 with a joystick is very incidental since my objective is initially to make it autonomous in the environment by avoiding going where it is not necessary like its little vacuum brother does and not to linger on the joysticks.
Thanks again for answering me
